README.md
# zora-cli

`zora-cli` is a cli tool for interacting with the Zora Protocol on the ethereum blockchain.

## usage

```
yarn global add @zoralabs/cli
```

### mint

#### uri 

Mint new zNFTs on the Zora Protocol from a contentURI and metadataURI.

##### local blockchain

```bash
zora mint uri --rpc-url http://localhost:8545 \
--wallet-mnemonic 'concert load couple harbor equip island argue ramp clarify fence smart topic' \
--chain-id 1337 \
--media-address 0x1D7022f5B17d2F8B695918FB48fa1089C9f85401 \
--market-address 0x1dC4c1cEFEF38a777b15aA20260a54E584b16C48 \
--content-uri https://ipfs.io/ipfs/bafybeiacyrrjel6qq2mdv6to6fvbzsg64y3g4kbp32h55zfmesx2oe7cwi \
--metadata-uri https://ipfs.io/ipfs/bafybeifvmzeyggdi36igqjiub5bq6vbpn5ebihxgqjbrk5ibqculk4upoy \
--creator-share 10 \
--timeout 10000
```

##### rinkeby

```bash
zora mint uri --rpc-url <infura or alchemy node> \
--wallet-mnemonic 'concert load couple harbor equip island argue ramp clarify fence smart topic' \
--chain-id 4 \
--content-uri https://ipfs.io/ipfs/bafybeiacyrrjel6qq2mdv6to6fvbzsg64y3g4kbp32h55zfmesx2oe7cwi \
--metadata-uri https://ipfs.io/ipfs/bafybeifvmzeyggdi36igqjiub5bq6vbpn5ebihxgqjbrk5ibqculk4upoy \
--creator-share 10 \
--timeout 10000
```

##### mainnet

```bash
zora mint uri --rpc-url <infura or alchemy node> \
--wallet-mnemonic 'concert load couple harbor equip island argue ramp clarify fence smart topic' \
--chain-id 1 \
--content-uri https://ipfs.io/ipfs/bafybeiacyrrjel6qq2mdv6to6fvbzsg64y3g4kbp32h55zfmesx2oe7cwi \
--metadata-uri https://ipfs.io/ipfs/bafybeifvmzeyggdi36igqjiub5bq6vbpn5ebihxgqjbrk5ibqculk4upoy \
--creator-share 10 \
--timeout 10000
```

#### file 

Coming Soon.

## Feature Requests

For feature requests please create github issues or post for discussion in our discord at https://zora.community.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- `zora mint file` to mint directly from a file on your computer
- `zora bid` to set, remove, and accept bids directly from cli
- `zora ask` to set and remove asks directly from cli

## [0.1.0] - 2021-02-19

### Added

- 'zora mint uri` cli tool.


src/commands/mint.ts
import { Command, command, metadata, option, Options } from 'clime'
import {
  constructMediaData,
  Zora,
  sha256FromBuffer,
  constructBidShares
} from '@zoralabs/zdk'
import { JsonRpcProvider } from '@ethersproject/providers'
import { Wallet } from 'ethers'
import axios from 'axios'

export class MintOptions extends Options {
  @option({
    flag: 'r',
    required: true,
    description: '[required] rpc endpoint for eth node'
  })
  rpcURL!: string

  @option({
    flag: 'w',
    required: true,
    description: '[required] wallet mnemonic'
  })
  walletMnemonic!: string

  @option({
    flag: 'i',
    required: true,
    description: '[required] chainId for the cli to connect to.'
    // TODO: validate chainID 1, 4, 50
  })
  chainId!: number

  @option({
    flag: 'c',
    required: true,
    description: '[required] uri where the content is located'
  })
  contentURI!: string

  @option({
    flag: 'm',
    required: true,
    description: '[required] uri where the metadata is located'
  })
  metadataURI!: string

  @option({
    flag: 's',
    required: true,
    description: "[required] creator's share (%) of future sales of the minted zNFT"
    // TODO: require that its less than 100
  })
  creatorShare!: number

  @option({
    flag: 'a',
    required: false,
    description: '[optional] override address for Zora Media Contract'
  })
  mediaAddress!: string

  @option({
    flag: 'b',
    required: false,
    description: '[optional] override address for Zora Market Contract'
  })
  marketAddress!: string

  @option({
    flag: 't',
    default: 10000,
    description: '[optional] timeout (ms) for making http requests to passed uris'
  })
  timeout!: number
}

@command({
  description:
    'This command will mint a new zNFT on the Zora Protocol. Use `zora mint --help` for more.'
})
export default class extends Command {
  @metadata
  async execute(options: MintOptions) {
    const provider = new JsonRpcProvider(options.rpcURL, options.chainId)
    await provider.ready

    const wallet = Wallet.fromMnemonic(options.walletMnemonic)
    const connectedWallet = wallet.connect(provider)

    let zora: Zora
    if (options.mediaAddress && options.marketAddress) {
      zora = new Zora(
        connectedWallet,
        options.chainId,
        options.mediaAddress,
        options.marketAddress
      )
    } else {
      zora = new Zora(connectedWallet, options.chainId)
    }

    const contentResp = await axios.get(options.contentURI, {
      timeout: options.timeout,
      responseType: 'arraybuffer'
    })
    const contentHash = sha256FromBuffer(contentResp.data)

    const metadataResp = await axios.get(options.metadataURI, {
      timeout: options.timeout,
      responseType: 'arraybuffer'
    })
    const metadataHash = sha256FromBuffer(metadataResp.data)

    const mediaData = constructMediaData(
      options.contentURI,
      options.metadataURI,
      contentHash,
      metadataHash
    )
    const bidShares = constructBidShares(
      options.creatorShare,
      100 - options.creatorShare,
      0
    )
    const tx = await zora.mint(mediaData, bidShares)

    console.log(
      `\n\n\tSuccessfully submitted Mint Transaction to the Zora Media Contract. \n\tContract Address: ${zora.mediaAddress}. \n\tChainId ${options.chainId}. \n\tTransaction Hash: ${tx.hash}.`
    )
  }
}


src/commands/mint/uri.ts
import { Command, command, metadata, option, Options } from 'clime'
import {
  constructMediaData,
  Zora,
  sha256FromBuffer,
  constructBidShares,
  isMediaDataVerified
} from '@zoralabs/zdk'
import { JsonRpcProvider } from '@ethersproject/providers'
import { Wallet } from 'ethers'
import axios from 'axios'

export class MintURIOptions extends Options {
  @option({
    flag: 'r',
    required: true,
    description: '[required] rpc endpoint for eth node'
  })
  rpcURL!: string

  @option({
    flag: 'w',
    required: true,
    description: '[required] wallet mnemonic'
  })
  walletMnemonic!: string

  @option({
    flag: 'i',
    required: true,
    description: '[required] chainId for the cli to connect to.'
    // TODO: validate chainID 1, 4, 50
  })
  chainId!: number

  @option({
    flag: 'c',
    required: true,
    description: '[required] uri where the content is located'
  })
  contentURI!: string

  @option({
    flag: 'm',
    required: true,
    description: '[required] uri where the metadata is located'
  })
  metadataURI!: string

  @option({
    flag: 's',
    required: true,
    description: "[required] creator's share (%) of future sales of the minted zNFT"
    // TODO: require that its less than 100
  })
  creatorShare!: number

  @option({
    flag: 'a',
    required: false,
    description: '[optional] override address for Zora Media Contract'
  })
  mediaAddress!: string

  @option({
    flag: 'b',
    required: false,
    description: '[optional] override address for Zora Market Contract'
  })
  marketAddress!: string

  @option({
    flag: 't',
    default: 10000,
    description: '[optional] timeout (ms) for making http requests to passed uris'
  })
  timeout!: number
}

@command({
  description:
    'This command will mint a new zNFT on the Zora Protocol from an existing contentURI and metadataURI. Type `zora mint uri --help` for more.'
})
export default class extends Command {
  @metadata
  async execute(options: MintURIOptions) {
    const provider = new JsonRpcProvider(options.rpcURL, options.chainId)
    await provider.ready

    const wallet = Wallet.fromMnemonic(options.walletMnemonic)
    const connectedWallet = wallet.connect(provider)

    let zora: Zora
    if (options.mediaAddress && options.marketAddress) {
      zora = new Zora(
        connectedWallet,
        options.chainId,
        options.mediaAddress,
        options.marketAddress
      )
    } else {
      zora = new Zora(connectedWallet, options.chainId)
    }

    const contentResp = await axios.get(options.contentURI, {
      timeout: options.timeout,
      responseType: 'arraybuffer'
    })
    const contentHash = sha256FromBuffer(contentResp.data)

    const metadataResp = await axios.get(options.metadataURI, {
      timeout: options.timeout,
      responseType: 'arraybuffer'
    })
    const metadataHash = sha256FromBuffer(metadataResp.data)

    const mediaData = constructMediaData(
      options.contentURI,
      options.metadataURI,
      contentHash,
      metadataHash
    )

    const bidShares = constructBidShares(
      options.creatorShare,
      100 - options.creatorShare,
      0
    )

    const verified = await isMediaDataVerified(mediaData, options.timeout)

    if (!verified){
      throw new Error("MediaData is not verified")
    }

    const tx = await zora.mint(mediaData, bidShares)

    console.log(
      `\n\nSuccessfully submitted Mint Transaction to the Zora Media Contract. \nContract Address: ${zora.mediaAddress}. \nChainId ${options.chainId}. \nTransaction Hash: ${tx.hash}.`
    )
  }
}
